src/Microsoft.DotNet.Tools.Publish/Program.cs

```diff
@@ -229,12 +230,55 @@ private static int PublishHost(ProjectContext context, string outputPath)
             return 0;
         }
 
-        private static void PublishFiles(IEnumerable<LibraryAsset> files, string outputPath)
+        private static void PublishFiles(IEnumerable<LibraryAsset> files, string outputPath, bool nativeSubdirectories)
         {
             foreach (var file in files)
             {
-                File.Copy(file.ResolvedPath, Path.Combine(outputPath, Path.GetFileName(file.ResolvedPath)), overwrite: true);
+                var destinationDirectory = DetermineFileDestinationDirectory(file, outputPath, nativeSubdirectories);
+
+                if (!Directory.Exists(destinationDirectory))
+                {
+                    Directory.CreateDirectory(destinationDirectory);
+                }
+
+                File.Copy(file.ResolvedPath, Path.Combine(destinationDirectory, Path.GetFileName(file.ResolvedPath)), overwrite: true);
+            }
+        }
+
+        private static string DetermineFileDestinationDirectory(LibraryAsset file, string outputPath, bool nativeSubdirectories)
+        {
+            var destinationDirectory = outputPath;
+
+            if (nativeSubdirectories)
+            {
+                destinationDirectory = Path.Combine(outputPath, GetNativeRelativeSubdirectory(file.RelativePath));
+            }
+
+            return destinationDirectory;
+        }
+
+        private static string GetNativeRelativeSubdirectory(string filepath)
+        {
+            string directoryPath = Path.GetDirectoryName(filepath);
+
+            string[] parts = directoryPath.Split(new string[] { "native" }, 2, StringSplitOptions.None);
+
+            if (parts.Length != 2)
+            {
+                throw new UnrecognizedNativeDirectoryFormat() { FailedPath = filepath };
             }
+
+            string candidate = parts[1];
+            candidate = candidate.TrimStart(new char[] { '/', '\\' });
+
+            return candidate;
+        }
+
+        private class UnrecognizedNativeDirectoryFormat : Exception
```

Comment:
I'm always a bit suspicious of custom `Exception`s. If this is truly exceptional, does a custom exception give you any useful information that couldn't be achieved through a message on an existing `Exception` type? If it isn't exceptional, we shouldn't use an exception for control flow.